File: src/dev/4.5/Legacy.EFLViewModelBase/BaseClasses/TypeDetector.vb
Imports System.Threading.Tasks
Imports System.Runtime.CompilerServices
Imports System.Reflection

Public Class TypeDetector

    Private myObjectToExamine As Object
    Private myTypes As List(Of Type)

    Private Sub New(objToTypeDetect As Object)
        myObjectToExamine = objToTypeDetect
    End Sub

    Private Sub FindTypes(t As Type)
        For Each propItem In myObjectToExamine.GetType.GetRuntimeProperties
            If propItem.PropertyType.GetTypeInfo.IsPrimitive Then
                Continue For
            End If

            If myTypes.Contains(propItem.PropertyType) Then
                Continue For
            Else
                myTypes.Add(propItem.PropertyType)
                FindTypes(propItem.PropertyType)
            End If
        Next
    End Sub

    Public Shared Function GetTypes(obj As Object) As IEnumerable(Of Type)

        If obj.GetType.GetTypeInfo.IsPrimitive Then
            Return New List(Of Type) From {obj.GetType}
        End If

        Dim td As New TypeDetector(obj)
        td.myTypes = New List(Of Type) From {obj.GetType}
        td.FindTypes(obj.GetType)
        Return td.myTypes

    End Function
End Class

File: src/dev/MvvmBaseLib/MvvmBaseLibShared/MvvmBaseClasses/TypeDetector.cs
using System;
using System.Collections.Generic;
using System.Reflection;

namespace ActiveDevelop.MvvmBaseLib.Mvvm
{
    public class TypeDetector
    {

        private object myObjectToExamine;
        private List<Type> myTypes;

        private TypeDetector(object objToTypeDetect)
        {
            myObjectToExamine = objToTypeDetect;
        }

        private void FindTypes(Type t)
        {
            foreach (var propItem in myObjectToExamine.GetType().GetRuntimeProperties())
            {
                if (propItem.PropertyType.GetTypeInfo().IsPrimitive)
                {
                    continue;
                }

                if (myTypes.Contains(propItem.PropertyType))
                {
                    continue;
                }
                else
                {
                    myTypes.Add(propItem.PropertyType);
                    FindTypes(propItem.PropertyType);
                }
            }
        }

        public static IEnumerable<Type> GetTypes(object obj)
        {

            if (obj.GetType().GetTypeInfo().IsPrimitive)
            {
                return new List<Type>() { obj.GetType() };
            }

            TypeDetector td = new TypeDetector(obj);
            td.myTypes = new List<Type>() { obj.GetType() };
            td.FindTypes(obj.GetType());
            return td.myTypes;

        }
    }
}